src/MongoDB.AspNetCore.OData/MongoEnableQueryAttribute.cs

```diff
@@ -0,0 +1,141 @@
+        request.HttpContext.Items[nameof(ODataQueryOptions)] = queryOptions;
+    }
+
+    public override void OnActionExecuted(ActionExecutedContext actionExecutedContext)
+    {
+        base.OnActionExecuted(actionExecutedContext);
+
+        var response = actionExecutedContext.HttpContext.Response;
+        // Check whether the response is set and successful
+        if (response != null && response.IsSuccessStatusCode() && actionExecutedContext.Result is ObjectResult responseContent)
+        {
+            // If response is IMongoQueryable it means default implementation bypasses all transformations,
+            // we want force apply projection to limit MQL request to struct properties only
+            if (responseContent.Value is IMongoQueryable mongoQueryable)
+            {
+                var queryOptions = (ODataQueryOptions)response.HttpContext.Items[nameof(ODataQueryOptions)];
+                var applyProjectionMethod = s_applyProjectionMethodInfo.MakeGenericMethod(mongoQueryable.ElementType);
+                responseContent.Value = applyProjectionMethod.Invoke(null, new object[] { mongoQueryable, queryOptions });
+            }
+        }
+    }
+
+    public override IQueryable ApplyQuery(IQueryable queryable, ODataQueryOptions queryOptions)
+    {
+        if (queryable is IMongoQueryable)
+        {
+            var applyImplMethodInfo = s_applyQueryInternalMethodInfo.MakeGenericMethod(queryable.ElementType);
+            return (IQueryable)applyImplMethodInfo.Invoke(null, new object[] { queryable, queryOptions });
+        }
+
+        return base.ApplyQuery(queryable, queryOptions);
+    }
+
+    private static IQueryable ApplyQueryInternal<T>(IQueryable<T> queryable, ODataQueryOptions queryOptions)
+    {
+        var ignoreQueryOptions = AllowedQueryOptions.Select | AllowedQueryOptions.Expand;
+        queryable = (IQueryable<T>)queryOptions.ApplyTo(queryable, ignoreQueryOptions);
+
+        if (queryOptions.Count?.Value == true)
+        {
+            return queryable;
+        }
+
+        queryable = ApplyProjection(queryable, queryOptions);
+        // SelectExpand projection generated by oData library is not compatible with Mongo LINQ provider,
+        // so we have to execute the query before applying it. And then reapply Select and Expand to in-memory objects.
+        // It should be done here to turn IMongoQueryable into IEnumerable, so the detection in OnActionExecuted works as expected.
+        queryable = ExecuteQuery(queryable);
+
+        if (queryOptions.SelectExpand != null)
+        {
+            var allowedOptions = ignoreQueryOptions | AllowedQueryOptions.Compute;
+            var ignoreOption = AllowedQueryOptions.All & ~allowedOptions;
+            return queryOptions.ApplyTo(queryable, ignoreOption);
+        }
+
+        return queryable;
+    }
+
+    private static IQueryable<T> ApplyProjection<T>(IQueryable<T> queryable, ODataQueryOptions queryOptions)
```

Comment:
Maybe receive `IMongoQueryable<T>`, as it's casted to anyway, and checked before invocation.

Reply:
Done